File: C-Cpp/src/t_crc32_ieee.cpp

#include "t_crc32_ieee.h"

//-------------------------------------------------------------------------------------------------------------------------------
//                                              CRC32 - IEEE
//-------------------------------------------------------------------------------------------------------------------------------
void crc32_make_table(uint32_t poly, uint32_t * table)
{
  for(uint32_t i = 0; i < 256; i++)
  {
    uint32_t byte = i;
    uint32_t crc = 0;
    for(uint32_t bit = 0; bit < 8; bit++)
    {
      if((byte ^ crc) & 1)
      {
        crc = (crc >> 1) ^ poly;
      }
      else
      {
        crc = crc >> 1;
      }
      byte = byte >> 1;
    }
    table[i] = crc;
  }
}

uint32_t crc32_calc(uint8_t * bytes, uint32_t len, uint32_t * table)
{
  uint32_t value = 0xffffffff;
  for(int i = 0; i < len; i++)
  {
    value = table[ (bytes[i] ^ value) & 0x000000ff ] ^ (value >> 8);
  }
  return (uint32_t)(-1 - (int32_t)value);
}

File: uPython/t_crc32_Ieee.py
#from a_data import data
from array import array
class crc32_ieee:
    #------------------------------------------------------------------------------------------------
    #                                              CRC32 - IEEE
    #------------------------------------------------------------------------------------------------
    
    table = 0
    rez = 0
    data_len = 0
    
    def make_table(poly):
        table = array('L')
        for byte in range(256):
            crc = 0
            for bit in range(8):
                if (byte ^ crc) & 1:
                    crc = (crc >> 1) ^ poly
                else:
                    crc >>= 1
                byte >>= 1
            table.append(crc)
        return table


    def calc(bytes, table):
        value = 0xffffffff
        for ch in bytes:
            value = table[(ch ^ value) & 0xff] ^ (value >> 8)
        return (-1 - value) & 0xFFFFFFFF
